### PR TITLE
Fix mSafari focus

### PR DESCRIPTION
**Pull request by staszekscp**
### Details

This PR fixes a bug of autofocus on mobile Safari, where keyboard was not opening, while navigating from a Modal opened via FAB.

cc: @ArekChr 

### Fixed Issues

$ https://github.com/Expensify/App/issues/10414
PROPOSAL: https://github.com/Expensify/App/issues/10414#issuecomment-1621710184

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Make sure the software keyboard is on - for simulator go to: `I/O` -> `Keyboard` and make sure the hardware keyboard is _**not**_ connected
2. Open mSafari
3. Press FAB (the green button with + in bottom right corner)
4. Open `New Chat`
5. See if the keyboard opens
6. Go back
7. Repeat steps 4-6 for `New Group` and `New Room` also

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/75020057/11892b20-31c8-4cbc-afd5-7dff5477f149

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/75020057/5e6cc2dc-88af-45d9-8399-7a1ce2c11f47

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/75020057/6b928a6b-4ecd-4967-8137-de38104abc48

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/75020057/72d0dd87-d50a-4b75-b5ba-f9b21fe5b2b5

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/75020057/8bfff51d-37d9-4ef5-9f10-ac8b3a995c70

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/75020057/7c176786-707e-4008-8847-f3022aa44a41

</details>

### CODE REVIEW

**Comment by ArekChr** [comment]
cc: [parasharrajat](https://github.com/parasharrajat) as C+ reviewer as well

File: src/pages/home/sidebar/SidebarScreen/index.js [hunk 1]
```diff
@@ -42,7 +46,9 @@ function SidebarScreen(props) {
             <BaseSidebarScreen
                 // eslint-disable-next-line react/jsx-props-no-spreading
                 {...props}
+                shouldShowOfflineIndicator={false}
             >
+                {props.isSmallScreenWidth && props.network.isOffline && <OfflineIndicator style={[StyleUtils.getBackgroundColorStyle(themeColors.greenHighlightBackground)]} />}
```
Comment (ArekChr) [inline src/pages/home/sidebar/SidebarScreen/index.js:51]: It seems the offline indicator is unrelated to this PR's purpose. Can you clarify if this is intentional?

File: src/components/Modal/index.web.js [hunk 2]
```diff
@@ -40,6 +44,20 @@ function Modal(props) {
...
+            coverScreen={false}
+            customBackdrop={
+                <Pressable
+                    onPress={(e) => {
+                        if (e && e.key === 'Enter') {
+                            return;
+                        }
+                        props.onClose();
+                    }}
+                    style={[styles.fullscreenFixed, {opacity: variables.overlayOpacity, backgroundColor: themeColors.overlay}]}
```
Comment (ArekChr) [inline src/components/Modal/index.web.js:56]: No inline styles. Keep it in styles files for maintainability and reusability.

File: src/components/Modal/modalPropTypes.js [hunk 3]
```diff
@@ -14,6 +14,12 @@ const propTypes = {
     /** Should we announce the Modal visibility changes? */
     shouldSetModalVisibility: PropTypes.bool,
 
+    /** Should we use coverScreen prop for ReactNativeModal? */
+    coverScreen: PropTypes.bool,
+
+    /** Custom backdrop JSX element */
+    customBackdrop: PropTypes.object,
```
Comment (ArekChr) [inline src/components/Modal/modalPropTypes.js:21]:
```suggestion
    customBackdrop: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
```

File: src/components/Modal/modalPropTypes.js [hunk 4]
```diff
@@ -14,6 +14,12 @@ const propTypes = {
     /** Should we announce the Modal visibility changes? */
     shouldSetModalVisibility: PropTypes.bool,
 
+    /** Should we use coverScreen prop for ReactNativeModal? */
+    coverScreen: PropTypes.bool,
+
+    /** Custom backdrop JSX element */
```
Comment (ArekChr) [inline src/components/Modal/modalPropTypes.js:20]:
```suggestion
    /** The customBackdrop prop allows the addition of a custom backdrop element behind the modal. */
```

File: src/components/PopoverMenu/BasePopoverMenu.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    ...createMenuPropTypes,
+    ...windowDimensionsPropTypes,
+
+    /** The horizontal and vertical anchors points for the popover */
+    anchorPosition: PropTypes.shape({
+        horizontal: PropTypes.number.isRequired,
+        vertical: PropTypes.number.isRequired,
+    }).isRequired,
+
+    /** Where the popover should be positioned relative to the anchor points. */
```
Comment (ArekChr) [inline src/components/PopoverMenu/BasePopoverMenu.js:26]:
```suggestion
    /** Sets the popover's position relative to the anchor points */
```

File: src/components/PopoverMenu/BasePopoverMenu.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import CONST from '../../CONST';
+import useArrowKeyFocusManager from '../../hooks/useArrowKeyFocusManager';
+import useKeyboardShortcut from '../../hooks/useKeyboardShortcut';
+import useWindowDimensions from '../../hooks/useWindowDimensions';
+
+const propTypes = {
+    ...createMenuPropTypes,
+    ...windowDimensionsPropTypes,
+
+    /** The horizontal and vertical anchors points for the popover */
```
Comment (ArekChr) [inline src/components/PopoverMenu/BasePopoverMenu.js:20]:
```suggestion
    /** Defines the anchor points for the popover */
```

File: src/components/PopoverMenu/BasePopoverMenu.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        vertical: PropTypes.number.isRequired,
+    }).isRequired,
+
+    /** Where the popover should be positioned relative to the anchor points. */
+    anchorAlignment: PropTypes.shape({
+        horizontal: PropTypes.oneOf(_.values(CONST.MODAL.ANCHOR_ORIGIN_HORIZONTAL)),
+        vertical: PropTypes.oneOf(_.values(CONST.MODAL.ANCHOR_ORIGIN_VERTICAL)),
+    }),
+
+    shouldNavigateBeforeClosingModal: PropTypes.bool,
```
Comment (ArekChr) [inline src/components/PopoverMenu/BasePopoverMenu.js:32]:
```suggestion
    /** Indicates whether navigation should occur before closing the modal */
    shouldNavigateBeforeClosingModal: PropTypes.bool,
```

File: src/components/Form.js [hunk 8]
```diff
@@ -466,6 +466,7 @@ export default compose(
     withOnyx({
         formState: {
             key: (props) => props.formID,
+            initWithStoredValues: false,
```
Comment (ArekChr) [inline src/components/Form.js:469]: Could you clarify why this change has been made? The form is used in many places across the codebase, and I'm not sure I understand why this modification has been made.

File: src/pages/home/sidebar/SidebarScreen/index.js (hunk 1 above)
Comment (staszekscp) [reply src/pages/home/sidebar/SidebarScreen/index.js:51]: Sure - the `coverScreen` prop set to false in `react-native-modal` changes the component from `Modal` to `View`. It caused the `OfflineIndicator` component to become visible, because it was adjacent to the `PopoverMenu`. I decided to slightly change the components' order to make sure that everything appears correctly on the screen, and the `PopoverMenu` stays on top.

File: src/components/Form.js (hunk 8 above)
Comment (staszekscp) [reply src/components/Form.js:472]: While I was working on this issue I noticed, that keyboard was not opening for Safari when some async calls from Onyx were executed. It was linked to reaching to the browser's local storage - especially when the call was taking ~50 ms. You are 100% right that it doesn't have to be changed for other platforms, so I'll use the `Browser.isMobileSafari()` instead. I'll also add explanation to make clear why it had to be added.

**Comment by staszekscp** [comment]
@parasharrajat Sorry, I hope it's ok now

File: src/components/PopoverMenu/BasePopoverMenu.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    /** Defines the anchor points for the popover */
+    anchorPosition: PropTypes.shape({
+        horizontal: PropTypes.number.isRequired,
+        vertical: PropTypes.number.isRequired,
+    }).isRequired,
+
+    /** Sets the popover's position relative to the anchor points */
+    anchorAlignment: PropTypes.shape({
+        horizontal: PropTypes.oneOf(_.values(CONST.MODAL.ANCHOR_ORIGIN_HORIZONTAL)),
+        vertical: PropTypes.oneOf(_.values(CONST.MODAL.ANCHOR_ORIGIN_VERTICAL)),
```
[the remark this answers is not in the document]
Comment (staszekscp) [reply src/components/PopoverMenu/BasePopoverMenu.js:29]: Hey! Thanks for the feedback. Actually this code was not written by me, I only moved this logic from `PopoverMenu.index.js` for reusability

**Comment by staszekscp** [comment]
Sorry, I'm out of office until Wednesday, I'll finish the PR then

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/pages/home/sidebar/SidebarScreen/index.js (hunk 1 above)
Comment (staszekscp) [reply src/pages/home/sidebar/SidebarScreen/index.js:51]: Unfortunately I was checking it, while I was writing this code, and a simple `zIndex` change will not work. The problem is most probably connected to the component tree and their `position` value, so when I was playing with it, it seemed that some bigger refactor would have to be applied in order to make it work with `zIndex`.

**Comment by staszekscp** [comment]
Thanks for the feedback, I've just applied the requested changes!

File: src/components/PopoverMenu/BasePopoverMenu.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    anchorAlignment: {
+        horizontal: CONST.MODAL.ANCHOR_ORIGIN_HORIZONTAL.LEFT,
+        vertical: CONST.MODAL.ANCHOR_ORIGIN_VERTICAL.BOTTOM,
+    },
+    shouldNavigateBeforeClosingModal: false,
+    anchorRef: () => {},
+    withoutOverlay: false,
+};
+
+function PopoverMenu(props) {
```
[the remark this answers is not in the document]
Comment (staszekscp) [reply src/components/PopoverMenu/BasePopoverMenu.js]: Sure, I have just changed it

**Comment by staszekscp** [comment]
Hey @parasharrajat! Have you had a chance to see if there is still something to be improved?

**Comment by staszekscp** [comment]
Hey! Sorry for the delay, I am currently focused on another task, which takes almost 100% of my time. Also, sorry to hear that you were feeling low, I hope you're better now!

As per your comment - Seeing the `Form` component I suspect that its refactor could help, so we wouldn't have to change the `initWithStoredValues` flag, but I cannot be 100% sure. In order to analyse the impact we would have to get access to the Safari code base which is not open-sourced. I understand your concern, however for now I don't see any alternative solution to make sure that opening keyboard would reliably open in each case. 

Therefore the question is - should we let it be merged, since there were no errors that were reproduced, and discovered so far, or should we maybe wait for the `Form` component refactor, and see if it helps in this edge case? I know that the work is in progress, but because it is a big component it may take a while.

Talking about the `OfflineIndicator` - I think that this change isn't an exception in the codebase, since there are already multiple cases when `shouldShowOfflineIndicator` prop is used. That's also why I decided to use this pattern :)

**Comment by staszekscp** [comment]
Hi! Because we're after the merge freeze I'd like to ask, should we proceed with this task?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by staszekscp** [comment]
Hey @parasharrajat, actually I noticed that something is not working after the newest main has been merged, so I have to investigate the issue, I'll let you know when I figure that out

File: src/components/Modal/index.web.js [hunk 11]
```diff
@@ -40,6 +43,15 @@ function Modal(props) {
...
             avoidKeyboard={false}
+            coverScreen={false}
+            customBackdrop={(onBackdropPress) => (
+                <Pressable
+                    onPress={onBackdropPress}
+                    style={styles.modalBackdropWeb}
+                    accessibilityLabel="backdrop"
+                    accessibilityRole={CONST.ACCESSIBILITY_ROLE.ADJUSTABLE}
+                />
+            )}
```
[the remark this answers is not in the document]
Comment (staszekscp) [reply src/components/Modal/index.web.js:54]: I don't think they are needed, since this is a fix only for mobile safari, we do not open keyboard on desktop, and the app works correctly there anyway

**Comment by staszekscp** [comment]
I made this solution work, but I noticed that the `initWithStoredValues` case has broaden a bit. After investigating the topic it turns out that the root cause of this edge case is `react-native-onyx`, and the fact that `withOnyx` while fetching data returns null. Apparently it is enough for Safari to not allow the keyboard opening. 

Setting `initWithStoredValues` to `false` for all the keys in the `Form` component "solves" the issue, but it seems to be a workaround anyway. I would like to have a closer look at that, but currently I'm involved in some high priority performance issues, so I won't be able to focus on this specific case early next week. On the other hand one of those issues is closely related to `react-native-onyx`, so I hope to get some more knowledge about the tool and come back with a solution (or at least an idea for it) soon. 😄 

I am sorry for the delays here. Unfortunately working with mSafari keyboard API is mostly guessing, since there is no clear explanation (or source code) available on how the mechanism of opening the keyboard exactly works. I think I got much understanding of it anyway, and I believe that fixing onyx is our last piece to make this merged.

**Comment by tgolen** [comment]
@staszekscp Let's keep up the work on this if you have some promising leads. You mentioned `initWithStoredValues` and there might be another recent change to `withOnyx()` that might be helpful. Check out [this PR](https://github.com/Expensify/react-native-onyx/pull/329). It's possible that by using `initialValue` or `delayUpdates` that it might help out the situation by letting the components render without waiting for data to be loaded from Onyx.

We have several experts when it comes to Onyx, so don't let your lack of knowledge on that lib hinder you! Start up a convo in Slack and we can help you out with any questions.

**Comment by staszekscp** [comment]
Hello! Recently I had a look again at this bug, and merged the newest `main`. Unfortunately it appears that my solution suddenly stopped working, and I have not been able to make it work again - regardless of changes in `react-native-onyx`. 

This PR has been here for a while, and even though it seemed that my solution worked (as you can see in videos), there was an edge case that I was able to solve only with a workaround. That's why I didn't want to push it to be merged, since I was not satisfied with the solution, anyway.

Right now I can see that my solution was fragile, and the bug could be easily introduced again. Currently I am focused on another task (wave-10), and I do not have much time to investigate the bug again. Therefore I don't want to block this task anymore and let someone alse to have a go and try to solve it... I am very sorry for it, and I hope someone else will be able to solve it. If that person has any questions I am available to share my knowledge about the subject. 😞 

Once again, I'm sorry 😢

**Comment by staszekscp** [comment]
Yes @parasharrajat, I'll do it. Also I'm going to close this PR. Sorry again 😞
